File: OfflinePlayback/src/main/java/com/bitmovin/player/samples/offline/playback/ListAdapter.java
/*
 * Bitmovin Player Android SDK
 * Copyright (C) 2017, Bitmovin GmbH, All Rights Reserved
 *
 * This source code and its use and distribution, is subject to the terms
 * and conditions of the applicable license agreement.
 */

package com.bitmovin.player.samples.offline.playback;

import android.content.Context;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageButton;
import android.widget.TextView;

import com.bitmovin.player.api.offline.options.OfflineContentOptions;
import com.bitmovin.player.api.offline.options.OfflineOptionEntry;
import com.bitmovin.player.api.offline.options.OfflineOptionEntryState;
import com.bitmovin.player.api.source.SourceConfig;

import java.util.List;

class ListAdapter extends ArrayAdapter<ListItem> {
    private ListItemActionListener listItemActionListener;

    public ListAdapter(@NonNull Context context, @LayoutRes int resource, @NonNull List<ListItem> objects, ListItemActionListener listItemActionListener) {
        super(context, resource, objects);
        this.listItemActionListener = listItemActionListener;
    }

    @NonNull
    @Override
    public View getView(final int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        View view = convertView;

        if (view == null) {
            view = LayoutInflater.from(getContext()).inflate(R.layout.list_item, parent, false);
        }
        final ListItem listItem = getItem(position);

        TextView title = view.findViewById(R.id.title);
        TextView state = view.findViewById(R.id.state);
        ImageButton pauseResume = view.findViewById(R.id.btnPauseResume);
        ImageButton btnDelete = view.findViewById(R.id.btnDelete);
        ImageButton btnDownload = view.findViewById(R.id.btnDownload);

        SourceConfig sourceConfig = listItem.getSourceConfig();
        OfflineContentOptions offlineContentOptions = listItem.getOfflineContentOptions();

        title.setText(sourceConfig.getTitle());
        if (offlineContentOptions != null) {
            btnDownload.setVisibility(View.VISIBLE);
            btnDownload.setOnClickListener(v -> listItemActionListener.showSelectionDialog(listItem));

            // If any option is downloading, we show a progress in the list
            if (isDownloading(offlineContentOptions)) {
                state.setText(String.format("Downloading - %.0f %%", listItem.getProgress()));
                state.setVisibility(View.VISIBLE);
                pauseResume.setImageResource(R.drawable.ic_pause_black_24dp);
                pauseResume.setVisibility(View.VISIBLE);
                pauseResume.setOnClickListener(v -> listItemActionListener.suspend(listItem));
            }
            else if (isSuspended(offlineContentOptions)) {
                pauseResume.setImageResource(R.drawable.ic_play_arrow_black_24dp);
                pauseResume.setVisibility(View.VISIBLE);
                pauseResume.setOnClickListener(v -> listItemActionListener.resume(listItem));
            }
            else if (hasFailed(offlineContentOptions)) {
                state.setText(String.format("Failed - %.0f %%", listItem.getProgress()));
                state.setVisibility(View.VISIBLE);
                pauseResume.setImageResource(R.drawable.ic_play_arrow_black_24dp);
                pauseResume.setVisibility(View.VISIBLE);
                pauseResume.setOnClickListener(v -> listItemActionListener.resume(listItem));
            }
            else {
                state.setVisibility(View.INVISIBLE);
                pauseResume.setVisibility(View.INVISIBLE);
            }
            // If any option is downloaded, we show the delete button
            if (hasDownloaded(offlineContentOptions)) {
                btnDelete.setVisibility(View.VISIBLE);
                btnDelete.setOnClickListener(v -> listItemActionListener.delete(listItem));
            }
            else {
                btnDelete.setVisibility(View.GONE);
            }

        }
        else {
            // If no options are available, we hide the download and the delete button
            btnDelete.setVisibility(View.GONE);
            btnDownload.setVisibility(View.GONE);
        }

        return view;
    }

    /**
     * Returns true, if one {@link OfflineOptionEntry}s state is {@link OfflineOptionEntryState#Downloading}
     *
     * @param offlineContentOptions
     * @return true, if one {@link OfflineOptionEntry}s state is {@link OfflineOptionEntryState#Downloading}
     */
    private boolean isDownloading(OfflineContentOptions offlineContentOptions) {
        List<OfflineOptionEntry> allOfflineOptionEntries = Util.getAsOneList(offlineContentOptions);
        for (OfflineOptionEntry entry : allOfflineOptionEntries) {
            if (entry.getState() == OfflineOptionEntryState.Downloading) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns true, if one {@link OfflineOptionEntry}s state is {@link OfflineOptionEntryState#Suspended}
     *
     * @param offlineContentOptions
     * @return true, if one {@link OfflineOptionEntry}s state is {@link OfflineOptionEntryState#Suspended}
     */
    private boolean isSuspended(OfflineContentOptions offlineContentOptions) {
        List<OfflineOptionEntry> allOfflineOptionEntries = Util.getAsOneList(offlineContentOptions);
        for (OfflineOptionEntry entry : allOfflineOptionEntries) {
            if (entry.getState() == OfflineOptionEntryState.Suspended) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns true, if one {@link OfflineOptionEntry}s state is {@link OfflineOptionEntryState#Failed}
     *
     * @param offlineContentOptions
     * @return true, if one {@link OfflineOptionEntry}s state is {@link OfflineOptionEntryState#Failed}
     */
    private boolean hasFailed(OfflineContentOptions offlineContentOptions) {
        List<OfflineOptionEntry> allOfflineOptionEntries = Util.getAsOneList(offlineContentOptions);
        for (OfflineOptionEntry entry : allOfflineOptionEntries) {
            if (entry.getState() == OfflineOptionEntryState.Failed) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns true, if one {@link OfflineOptionEntry}s state is {@link OfflineOptionEntryState#Downloaded}
     *
     * @param offlineContentOptions
     * @return true, if one {@link OfflineOptionEntry}s state is {@link OfflineOptionEntryState#Downloaded}
     */
    private boolean hasDownloaded(OfflineContentOptions offlineContentOptions) {
        List<OfflineOptionEntry> allOfflineOptionEntries = Util.getAsOneList(offlineContentOptions);
        for (OfflineOptionEntry entry : allOfflineOptionEntries) {
            if (entry.getState() == OfflineOptionEntryState.Downloaded) {
                return true;
            }
        }
        return false;
    }
}

File: OfflinePlaybackKotlin/src/main/java/com/bitmovin/player/samples/offline/playback/ListAdapter.kt
package com.bitmovin.player.samples.offline.playback

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.annotation.LayoutRes
import com.bitmovin.player.api.offline.options.OfflineContentOptions
import com.bitmovin.player.api.offline.options.OfflineOptionEntryState
import com.bitmovin.player.samples.offline.playback.databinding.ListItemBinding

class ListAdapter(
        context: Context,
        @LayoutRes resource: Int,
        objects: List<ListItem>,
        private val listItemActionListener: ListItemActionListener
) : ArrayAdapter<ListItem>(context, resource, objects) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val binding = convertView?.let { ListItemBinding.bind(convertView) }
            ?: ListItemBinding.inflate(LayoutInflater.from(context), parent, false)

        val listItem = getItem(position)

        val sourceConfig = listItem?.sourceConfig
        val offlineContentOptions = listItem?.offlineContentOptions
        binding.title.text = sourceConfig?.title

        if (offlineContentOptions != null) {
            binding.btnDownload.visibility = View.VISIBLE
            binding.btnDownload.setOnClickListener { listItemActionListener.showSelectionDialog(listItem) }

            // If any option is downloading, show a progress in the list
            when {
                isDownloading(offlineContentOptions) -> {
                    binding.apply {
                        state.text = String.format("Downloading - %.0f %%", listItem.progress)
                        state.visibility = View.VISIBLE
                        btnPauseResume.setImageResource(R.drawable.ic_pause_black_24dp)
                        btnPauseResume.visibility = View.VISIBLE
                        btnPauseResume.setOnClickListener { listItemActionListener.suspend(listItem) }
                    }
                }
                isSuspended(offlineContentOptions) -> {
                    binding.btnPauseResume.apply {
                        setImageResource(R.drawable.ic_play_arrow_black_24dp)
                        visibility = View.VISIBLE
                        setOnClickListener { listItemActionListener.resume(listItem) }
                    }
                }
                hasFailed(offlineContentOptions) -> {
                    binding.apply {
                        state.text = String.format("Failed - %.0f %%", listItem.progress)
                        state.visibility = View.VISIBLE
                        btnPauseResume.setImageResource(R.drawable.ic_play_arrow_black_24dp)
                        btnPauseResume.visibility = View.VISIBLE
                        btnPauseResume.setOnClickListener { listItemActionListener.resume(listItem) }
                    }
                }
                else -> {
                    binding.state.visibility = View.INVISIBLE
                    binding.btnPauseResume.visibility = View.INVISIBLE
                }
            }
            // If any option is downloaded, we show the delete button
            if (hasDownloaded(offlineContentOptions)) {
                binding.btnDelete.visibility = View.VISIBLE
                binding.btnDelete.setOnClickListener { listItemActionListener.delete(listItem) }
            } else {
                binding.btnDelete.visibility = View.GONE
            }

        } else {
            // If no options are available, we hide the download and the delete button
            binding.btnDelete.visibility = View.GONE
            binding.btnDownload.visibility = View.GONE
        }

        return binding.root
    }

    /**
     * Returns true, if one [OfflineOptionEntry]s state is [OfflineOptionEntryState.Downloading]
     *
     * @param offlineContentOptions
     * @return true, if one [OfflineOptionEntry]s state is [OfflineOptionEntryState.Downloading]
     */
    private fun isDownloading(offlineContentOptions: OfflineContentOptions): Boolean {
        val allOfflineOptionEntries = Util.getAsOneList(offlineContentOptions)
        return allOfflineOptionEntries.any { entry -> entry.state == OfflineOptionEntryState.Downloading }
    }

    /**
     * Returns true, if one [OfflineOptionEntry]s state is [OfflineOptionEntryState.Suspended]
     *
     * @param offlineContentOptions
     * @return true, if one [OfflineOptionEntry]s state is [OfflineOptionEntryState.Suspended]
     */
    private fun isSuspended(offlineContentOptions: OfflineContentOptions): Boolean {
        val allOfflineOptionEntries = Util.getAsOneList(offlineContentOptions)
        return allOfflineOptionEntries.any { entry -> entry.state == OfflineOptionEntryState.Suspended }
    }

    /**
     * Returns true, if one [OfflineOptionEntry]s state is [OfflineOptionEntryState.Failed]
     *
     * @param offlineContentOptions
     * @return true, if one [OfflineOptionEntry]s state is [OfflineOptionEntryState.Failed]
     */
    private fun hasFailed(offlineContentOptions: OfflineContentOptions): Boolean {
        val allOfflineOptionEntries = Util.getAsOneList(offlineContentOptions)
        return allOfflineOptionEntries.any { entry -> entry.state == OfflineOptionEntryState.Failed }
    }

    /**
     * Returns true, if one [OfflineOptionEntry]s state is [OfflineOptionEntryState.Downloaded]
     *
     * @param offlineContentOptions
     * @return true, if one [OfflineOptionEntry]s state is [OfflineOptionEntryState.Downloaded]
     */
    private fun hasDownloaded(offlineContentOptions: OfflineContentOptions): Boolean {
        val allOfflineOptionEntries = Util.getAsOneList(offlineContentOptions)
        return allOfflineOptionEntries.any { entry -> entry.state == OfflineOptionEntryState.Downloaded }
    }
}
